runtime/gc_glue_java/EnvironmentDelegate.cpp

```diff
@@ -252,58 +252,134 @@ MM_EnvironmentDelegate::forceOutOfLineVMAccess()
 }
 
 /**
- * Re-enable inline TLH allocate by restoring heapAlloc from realHeapAlloc
+ * Re-enable inline TLH allocate by restoring heapTop from realHeapTop
  */
 void
 MM_EnvironmentDelegate::enableInlineTLHAllocate()
 {
 	J9ModronThreadLocalHeap *tlh = (J9ModronThreadLocalHeap *)&_vmThread->allocateThreadLocalHeap;
-	_vmThread->heapAlloc =  tlh->realHeapAlloc;
-	tlh->realHeapAlloc = NULL;
+	if (NULL != tlh->realHeapTop) {
+		_vmThread->heapTop =  tlh->realHeapTop;
+		tlh->realHeapTop = NULL;
+	}
 
 #if defined(J9VM_GC_NON_ZERO_TLH)
 	tlh = (J9ModronThreadLocalHeap *)&_vmThread->nonZeroAllocateThreadLocalHeap;
-	_vmThread->nonZeroHeapAlloc =  tlh->realHeapAlloc;
-	tlh->realHeapAlloc = NULL;
+	if (NULL != tlh->realHeapTop) {
+		_vmThread->nonZeroHeapTop =  tlh->realHeapTop;
+		tlh->realHeapTop = NULL;
+	}
 #endif /* defined(J9VM_GC_NON_ZERO_TLH) */
 }
 
 /**
- * Determine if inline TLH allocate is enabled; its enabled if realheapAlloc is NULL.
+ * Determine if inline TLH allocate is enabled; its enabled if realheapTop is NULL.
  * @return TRUE if inline TLH allocates currently enabled for this thread; FALSE otherwise
  */
 bool
 MM_EnvironmentDelegate::isInlineTLHAllocateEnabled()
 {
 	J9ModronThreadLocalHeap *tlh = (J9ModronThreadLocalHeap *)&_vmThread->allocateThreadLocalHeap;
-	bool result = (NULL == tlh->realHeapAlloc);
+	bool result = (NULL == tlh->realHeapTop);
 
 #if defined(J9VM_GC_NON_ZERO_TLH)
 	tlh = (J9ModronThreadLocalHeap *)&_vmThread->nonZeroAllocateThreadLocalHeap;
-	result = result && (NULL == tlh->realHeapAlloc);
+	result = result && (NULL == tlh->realHeapTop);
 #endif /* defined(J9VM_GC_NON_ZERO_TLH) */
 
 	return result;
 }
+
+
+void
+MM_EnvironmentDelegate::setTLHSamplingTop(uintptr_t size)
+{
+	uintptr_t max = 0;
+	J9ModronThreadLocalHeap *tlh = (J9ModronThreadLocalHeap *)&_vmThread->allocateThreadLocalHeap;
+	if (NULL != tlh->realHeapTop) {
+		max = tlh->realHeapTop - _vmThread->heapAlloc;
+	} else {
+		max = _vmThread->heapTop - _vmThread->heapAlloc;
+	}
+
+	if (max > size) {
+		if (NULL == tlh->realHeapTop) {
+			tlh->realHeapTop = _vmThread->heapTop;
+		}
+		_vmThread->heapTop = _vmThread->heapAlloc + size;
+	} else {
+		if (NULL != tlh->realHeapTop) {
+			_vmThread->heapTop =  tlh->realHeapTop;
+			tlh->realHeapTop = NULL;
+		}
+	}
+
+#if defined(J9VM_GC_NON_ZERO_TLH)
+	/* TODO: need to find better way set TLH sampling Top for NON ZERO case */
+	tlh = (J9ModronThreadLocalHeap *)&_vmThread->nonZeroAllocateThreadLocalHeap;
+	if (NULL != tlh->realHeapTop) {
+		max = tlh->realHeapTop - _vmThread->nonZeroHeapAlloc;
+	} else {
+		max = _vmThread->nonZeroHeapTop - _vmThread->nonZeroHeapAlloc;
+	}
+
+	if (max > size) {
+		if (NULL == tlh->realHeapTop) {
+			tlh->realHeapTop = _vmThread->nonZeroHeapTop;
+		}
+		_vmThread->nonZeroHeapTop = _vmThread->nonZeroHeapAlloc + size;
+	} else {
+		if (NULL != tlh->realHeapTop) {
+			_vmThread->nonZeroHeapTop =  tlh->realHeapTop;
+			tlh->realHeapTop = NULL;
+		}
+	}
```

Comment:
Looks like the question I asked about support for dual TLH mode applies here: Am I reading correctly that if we set size for both TLHs we can get 2 * size to be allocated potentially?

Reply:
yes, we might not get very accurate result on nonzero case,have not found the better way to handle the non zero case.

Reply:
Do we have a platform that can actually have both active (btw, is X using just nonZeroHeapAlloc/Top?)?  I'm willing to ignore that issue for now, if we think it will take more than a day to do/test it and we can follow up after the upcoming release (by as you said disabling one of two TLHs).

Reply:
I believe non-Zeroed TLH is using on pLinux (LE and BE) and AIX. I am not sure about current status of zLinux.

Reply:
Non-zero TLH can be used for primitive arrays only. So to see a mismatch an application should include primitive arrays to the allocation mix.

Reply:
I agree that proper handling for dual TLH case can be done later in separate change